yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,237 @@
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _initialize_state_variables(self):
+        super()._initialize_state_variables()
+        self.num_grids = 1
+
+    def _detect_output_fields(self):
+        # This needs to set a self.field_list that contains all the available,
+        # on-disk fields. No derived fields should be defined here.
+        # NOTE: Each should be a tuple, where the first element is the on-disk
+        # fluid type or particle type.  Convention suggests that the on-disk
+        # fluid type is usually the dataset_type and the on-disk particle type
+        # (for a single population of particles) is "io".
+        self.field_list = []
+        for key in self.ds._handle.variables.keys():
+            if (
+                all(x in self.ds._handle[key].dims for x in ["time", "z", "y", "x"])
+                is True
+            ):
+                field_tup = ("cf_radial", key)
+                self.field_list.append(field_tup)
+
+    def _count_grids(self):
+        # This needs to set self.num_grids
+        pass
+
+    def _parse_index(self):
+        # This needs to fill the following arrays, where N is self.num_grids:
+        #   self.grid_left_edge         (N, 3) <= float64
+        #   self.grid_right_edge        (N, 3) <= float64
+        #   self.grid_dimensions        (N, 3) <= int
+        #   self.grid_particle_count    (N, 1) <= int
+        #   self.grid_levels            (N, 1) <= int
+        #   self.grids                  (N, 1) <= grid objects
+        #   self.max_level = self.grid_levels.max()
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grids = np.empty(self.num_grids, dtype="object")
+        for i in range(self.num_grids):
+            g = self.grid(i, self, self.grid_levels.flat[i], self.grid_dimensions[i])
+            g._prepare_grid()
+            g._setup_dx()
+            self.grids[i] = g
+        self.max_level = 1
+        pass
+
+    def _populate_grid_objects(self):
+        # For each grid, this must call:
+        #   grid._prepare_grid()
+        #   grid._setup_dx()
+        # This must also set:
+        #   grid.Children <= list of child grids
+        #   grid.Parent   <= parent grid
+        # This is handled by the frontend because often the children must be
+        # identified.
+        pass
+
+
+class CFRadialDataset(Dataset):
+    _index_class = CFRadialHierarchy
+    _field_info_class = CFRadialFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="cf_radial",
+        storage_filename=None,
+        units_override=None,
+    ):
+        self.fluid_types += ("cf_radial",)
+        self._handle = xarray.open_dataset(filename)
+        self.refine_by = 2
+        if "x" not in self._handle.coords:
```

Comment:
Could add a check for the existence of the gridded file here in case we've already gridded and saved the new file in a previous load (unless you see a reason to re-grid every time).

Reply:
For this, I believe when we wrote this, the if x was the check if the file was a grid file, if so we skip the pyart gridding to cartesian step. I can also add to the if: and "_grid.nc" not in filename, but wasn't too sure and state our original intentions might have covered this one.

Reply:
ok, that makes some sense. My point is just that if a user calls `yt.load("cfradial_notgridded.nc")`, it would correctly trigger the pyart gridding, after which "cfradial_notgridded_gridded.nc" would exist. So in subsequent calls to `yt.load("cfradial_notgridded.nc")` it'd be nice to simply reload "cfradial_notgridded_gridded.nc" since it already exists. I was thinking something like (in pseudo-code):

```python
if "x" not in self._handle.coords:

    if gridded_file does not exist:
         do the pyart gridding, create gridded file

    load up the gridded file into self._handle    
```

This really only matters if the gridding step is computationally difficult -- if it's quick then doing it every time is OK, but if the gridding step takes some time, why not re-use the gridded files?

Reply:
It depends on the grid dimensions, so it could be slow, so I agree. I added a check, I believe I did it as you mentioned, if not I can fix it if need be.